In [ ]:
from google.colab import files

uploaded = files.upload()

Saving links.csv to links.csv
Saving movies.csv to movies.csv
Saving ratings.csv to ratings.csv
Saving tags.csv to tags.csv


In [ ]:
import pandas as pd
import numpy as np

movies = pd.read_csv("movies.csv")
ratings = pd.read_csv("ratings.csv")
tags = pd.read_csv("tags.csv")
links = pd.read_csv("links.csv")

print("Movies shape:", movies.shape)
print("Ratings shape:", ratings.shape)
print("Tags shape:", tags.shape)
print("Links shape:", links.shape)

Movies shape: (9742, 3)
Ratings shape: (100836, 4)
Tags shape: (3683, 4)
Links shape: (9742, 3)


In [ ]:
print("MOVIES DATASET")
display(movies.head())

print("\nRATINGS DATASET")
display(ratings.head())

print("\nTAGS DATASET")
display(tags.head())

print("\nLINKS DATASET")
display(links.head())

MOVIES DATASET


,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy



RATINGS DATASET


,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931



TAGS DATASET


,userId,movieId,tag,timestamp
0,2,60756,funny,1445714994
1,2,60756,Highly quotable,1445714996
2,2,60756,will ferrell,1445714992
3,2,89774,Boxing story,1445715207
4,2,89774,MMA,1445715200



LINKS DATASET


,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


In [ ]:
# Check basic information about each dataset

datasets = {
    "Movies": movies,
    "Ratings": ratings,
    "Tags": tags,
    "Links": links
}

for name, df in datasets.items():
    print("=" * 50)
    print(name.upper(), "DATASET")
    print("=" * 50)

    print("\nShape:", df.shape)

    print("\nData Types:")
    print(df.dtypes)

    print("\nMissing Values:")
    print(df.isnull().sum())

    print("\nDuplicate Rows:", df.duplicated().sum())

    print()

MOVIES DATASET

Shape: (9742, 3)

Data Types:
movieId     int64
title      object
genres     object
dtype: object

Missing Values:
movieId    0
title      0
genres     0
dtype: int64

Duplicate Rows: 0

RATINGS DATASET

Shape: (100836, 4)

Data Types:
userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object

Missing Values:
userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

Duplicate Rows: 0

TAGS DATASET

Shape: (3683, 4)

Data Types:
userId        int64
movieId       int64
tag          object
timestamp     int64
dtype: object

Missing Values:
userId       0
movieId      0
tag          0
timestamp    0
dtype: int64

Duplicate Rows: 0

LINKS DATASET

Shape: (9742, 3)

Data Types:
movieId      int64
imdbId       int64
tmdbId     float64
dtype: object

Missing Values:
movieId    0
imdbId     0
tmdbId     8
dtype: int64

Duplicate Rows: 0



In [ ]:
print("Number of unique users:", ratings["userId"].nunique())
print("Number of unique rated movies:", ratings["movieId"].nunique())

print("\nRating distribution:")
print(ratings["rating"].value_counts().sort_index())

print("\nRating statistics:")
display(ratings["rating"].describe())

Number of unique users: 610
Number of unique rated movies: 9724

Rating distribution:
rating
0.5     1370
1.0     2811
1.5     1791
2.0     7551
2.5     5550
3.0    20047
3.5    13136
4.0    26818
4.5     8551
5.0    13211
Name: count, dtype: int64

Rating statistics:


,rating
count,100836.000000
mean,3.501557
std,1.042529
min,0.500000
25%,3.000000
50%,3.500000
75%,4.000000
max,5.000000


In [ ]:
# Create copies to preserve original data
movies_clean = movies.copy()
ratings_clean = ratings.copy()
tags_clean = tags.copy()
links_clean = links.copy()

# Extract release year from movie title
movies_clean["year"] = movies_clean["title"].str.extract(r"\((\d{4})\)$")

# Remove year from title to create a clean movie title
movies_clean["clean_title"] = movies_clean["title"].str.replace(
    r"\s*\(\d{4}\)$", "", regex=True
)

# Convert year to numeric
movies_clean["year"] = pd.to_numeric(
    movies_clean["year"], errors="coerce"
)

# Convert Unix timestamps to readable datetime
ratings_clean["datetime"] = pd.to_datetime(
    ratings_clean["timestamp"], unit="s"
)

tags_clean["datetime"] = pd.to_datetime(
    tags_clean["timestamp"], unit="s"
)

# Preview processed data
print("PROCESSED MOVIES")
display(movies_clean.head())

print("\nPROCESSED RATINGS")
display(ratings_clean.head())

print("\nPROCESSED TAGS")
display(tags_clean.head())

PROCESSED MOVIES


,movieId,title,genres,year,clean_title
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,1995.0,Toy Story
1,2,Jumanji (1995),Adventure|Children|Fantasy,1995.0,Jumanji
2,3,Grumpier Old Men (1995),Comedy|Romance,1995.0,Grumpier Old Men
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,1995.0,Waiting to Exhale
4,5,Father of the Bride Part II (1995),Comedy,1995.0,Father of the Bride Part II



PROCESSED RATINGS


,userId,movieId,rating,timestamp,datetime
0,1,1,4.0,964982703,2000-07-30 18:45:03
1,1,3,4.0,964981247,2000-07-30 18:20:47
2,1,6,4.0,964982224,2000-07-30 18:37:04
3,1,47,5.0,964983815,2000-07-30 19:03:35
4,1,50,5.0,964982931,2000-07-30 18:48:51



PROCESSED TAGS


,userId,movieId,tag,timestamp,datetime
0,2,60756,funny,1445714994,2015-10-24 19:29:54
1,2,60756,Highly quotable,1445714996,2015-10-24 19:29:56
2,2,60756,will ferrell,1445714992,2015-10-24 19:29:52
3,2,89774,Boxing story,1445715207,2015-10-24 19:33:27
4,2,89774,MMA,1445715200,2015-10-24 19:33:20


In [ ]:
# Split genres into Python lists
movies_clean["genre_list"] = movies_clean["genres"].str.split("|")

# Create a lowercase version of tags for consistent NLP processing
tags_clean["tag_clean"] = (
    tags_clean["tag"]
    .str.lower()
    .str.strip()
)

# Combine all tags belonging to the same movie
movie_tags = (
    tags_clean.groupby("movieId")["tag_clean"]
    .apply(lambda x: " ".join(x))
    .reset_index()
)

movie_tags.rename(columns={"tag_clean": "combined_tags"}, inplace=True)

# Merge combined tags with movie information
movies_enriched = movies_clean.merge(
    movie_tags,
    on="movieId",
    how="left"
)

# Movies without user tags get an empty string
movies_enriched["combined_tags"] = movies_enriched["combined_tags"].fillna("")

print("ENRICHED MOVIE DATA")
display(
    movies_enriched[
        ["movieId", "clean_title", "year", "genres",
         "genre_list", "combined_tags"]
    ].head(10)
)

print("\nShape:", movies_enriched.shape)

ENRICHED MOVIE DATA


,movieId,clean_title,year,genres,genre_list,combined_tags
0,1,Toy Story,1995.0,Adventure|Animation|Children|Comedy|Fantasy,"[Adventure, Animation, Children, Comedy, Fantasy]",pixar pixar fun
1,2,Jumanji,1995.0,Adventure|Children|Fantasy,"[Adventure, Children, Fantasy]",fantasy magic board game robin williams game
2,3,Grumpier Old Men,1995.0,Comedy|Romance,"[Comedy, Romance]",moldy old
3,4,Waiting to Exhale,1995.0,Comedy|Drama|Romance,"[Comedy, Drama, Romance]",
4,5,Father of the Bride Part II,1995.0,Comedy,[Comedy],pregnancy remake
5,6,Heat,1995.0,Action|Crime|Thriller,"[Action, Crime, Thriller]",
6,7,Sabrina,1995.0,Comedy|Romance,"[Comedy, Romance]",remake
7,8,Tom and Huck,1995.0,Adventure|Children,"[Adventure, Children]",
8,9,Sudden Death,1995.0,Action,[Action],
9,10,GoldenEye,1995.0,Action|Adventure|Thriller,"[Action, Adventure, Thriller]",



Shape: (9742, 7)


In [ ]:
# Calculate rating statistics for each movie
rating_stats = (
    ratings_clean.groupby("movieId")
    .agg(
        rating_count=("rating", "count"),
        average_rating=("rating", "mean")
    )
    .reset_index()
)

# Merge rating statistics with enriched movie data
movies_final = movies_enriched.merge(
    rating_stats,
    on="movieId",
    how="left"
)

# Fill missing values for movies with no ratings
movies_final["rating_count"] = movies_final["rating_count"].fillna(0).astype(int)
movies_final["average_rating"] = movies_final["average_rating"].fillna(0)

# Round average rating
movies_final["average_rating"] = movies_final["average_rating"].round(2)

print("FINAL MOVIE DATASET")
display(
    movies_final[
        [
            "movieId",
            "clean_title",
            "year",
            "genres",
            "combined_tags",
            "rating_count",
            "average_rating"
        ]
    ].head(10)
)

print("\nFinal Shape:", movies_final.shape)

FINAL MOVIE DATASET


,movieId,clean_title,year,genres,combined_tags,rating_count,average_rating
0,1,Toy Story,1995.0,Adventure|Animation|Children|Comedy|Fantasy,pixar pixar fun,215,3.92
1,2,Jumanji,1995.0,Adventure|Children|Fantasy,fantasy magic board game robin williams game,110,3.43
2,3,Grumpier Old Men,1995.0,Comedy|Romance,moldy old,52,3.26
3,4,Waiting to Exhale,1995.0,Comedy|Drama|Romance,,7,2.36
4,5,Father of the Bride Part II,1995.0,Comedy,pregnancy remake,49,3.07
5,6,Heat,1995.0,Action|Crime|Thriller,,102,3.95
6,7,Sabrina,1995.0,Comedy|Romance,remake,54,3.19
7,8,Tom and Huck,1995.0,Adventure|Children,,8,2.88
8,9,Sudden Death,1995.0,Action,,16,3.12
9,10,GoldenEye,1995.0,Action|Adventure|Thriller,,132,3.50



Final Shape: (9742, 9)


In [ ]:
import os

# Create output folder
os.makedirs("processed_data", exist_ok=True)

# Save processed datasets
movies_final.to_csv(
    "processed_data/movies_final.csv",
    index=False
)

ratings_clean.to_csv(
    "processed_data/ratings_clean.csv",
    index=False
)

tags_clean.to_csv(
    "processed_data/tags_clean.csv",
    index=False
)

print("Processed datasets saved successfully!")

Processed datasets saved successfully!


## Data Understanding – Key Findings

- Dataset contains **9,742 movies**, **100,836 ratings**, **610 unique users**, and **3,683 tags**.
- No duplicate records were found, and the datasets are mostly free from missing values.
- User ratings range from **0.5 to 5.0**, with an average rating of approximately **3.5**.
- Movie titles were cleaned and release years were extracted.
- Genres were processed and user-generated tags were combined for each movie.
- Rating statistics such as **rating count** and **average rating** were calculated for every movie.
- A final enriched movie dataset of **9,742 movies with 9 features** was created, ready for further analysis and recommendation system development.